### DISCUSSION TITLE
State or image switching when the button is pressed

### DISCUSSION THREAD

**Discussion by pyking-wang** [AvaloniaUI/Avalonia#11910] [Q&A] [answered]
I have a wheel button and I want to switch its display image when pressed. What is the best way to do it, using a button or toggle button?
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/4ca61475-c584-4bf3-991d-db3f54987d75)
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/bde761e9-fd58-413c-88c7-54c7c0dce212)
Figure 1 shows the normal state, while Figure 2 only shows the state when pressed down

**Comment by pyking-wang** [comment]
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/bdc7227f-0f7b-4f0c-9bc1-11196a5ffb04)
By the way, is the official document example no longer available?

**Reply by timunie** [reply]
We are working heavily on updating the docs. So atm there may be some inconvinience.

**Comment by timunie** [comment]
> I have a wheel button and I want to switch its display image when pressed. What is the best way to do it, using a button or toggle button? ![image](https://user-images.githubusercontent.com/76192669/249056779-4ca61475-c584-4bf3-991d-db3f54987d75.png) ![image](https://user-images.githubusercontent.com/76192669/249056931-bde761e9-fd58-413c-88c7-54c7c0dce212.png) Figure 1 shows the normal state, while Figure 2 only shows the state when pressed down

Well, depends on your needs. If I can bind it to a bool in the ViewModel, I'd use a `ToggleButton`. if it should really only be changed while the `Button` is pressed, use a `Button`. The other question for me is, do you need to use an image? from what I see above, a `Path` or `PathIcon` (v11 only) which can change the fill seems to be sufficient. 

Here is an example using Path: 

```xml
<Window.Styles>
    <!-- Add a style for normal state.  -->
    <Style Selector="Button.img > Path">
        <Setter Property="Fill" Value="Gray"></Setter>
        <Setter Property="Stretch" Value="UniformToFill"></Setter>
        <Setter Property="Width" Value="20"></Setter>
        <Setter Property="Height" Value="20"></Setter>
    </Style>

    <!-- Add a style for pressed state -->
    <Style Selector="Button.img:pressed > Path">
        <Setter Property="Fill" Value="DodgerBlue"></Setter>
    </Style>
</Window.Styles>

<Button Classes="img">
    <Path Data="M 9.2604166,23.8125 11.90625,21.166666 14.552083,23.8125 h -1.322916 v 3.96875 H 10.583333 V 23.8125 Z" />
</Button>
```

**Reply by pyking-wang** [reply]
Thank you for your reply. I am currently porting some projects that use images in different states. Thank you very much for giving me ideas in this way. In order to reduce workload, it seems that I still need to use<Button><Image/></Button>.

**Reply by timunie** [reply]
is the state a bool on your ViewModel? if so, just add two images as content and bind visibility to that bool.

**Reply by pyking-wang** [reply]
Binding Bool, I think it should only be used for Click.This is feasible.
If it's Button: pressed, what should I do with two images？

**Reply by timunie** [reply, marked as answer]
maybe like that with named images:

```xml
<!-- normal state -->
<Style Selector="Button.img Image#NormalImage">
    <Setter Property="IsVisible" Value="True" />
</Style>

<Style Selector="Button.img Image#PressedImage">
    <Setter Property="IsVisible" Value="False" />
</Style>

<!-- pressed state -->
<Style Selector="Button.img:pressed Image#NormalImage">
    <Setter Property="IsVisible" Value="False" />
</Style>

<Style Selector="Button.img:pressed Image#PressedImage">
    <Setter Property="IsVisible" Value="True" />
</Style>
```

usage

```xml
<Button Classes="img" HorizontalAlignment="Center">
    <Panel>
        <Image x:Name="NormalImage" Source="/Assets/sad.png" Stretch="None" />
        <Image x:Name="PressedImage" Source="/Assets/happy.png" Stretch="None" />
    </Panel>
</Button>
```

![Animation](https://github.com/AvaloniaUI/Avalonia/assets/47110241/25ba6168-95d4-417c-aba9-b2c8d1a714e3)

**Reply by pyking-wang** [reply]
This is the effect I need, thank you very much! 
I will close it. 
By the way, is it necessary to declare image resources like this, and does it affect performance?
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/557c43ee-29b7-4098-87ae-ff41329cf7b0)

**Reply by timunie** [reply]
<Bitmap x:Key="MyKey" Source="/Assets/MyImag.png" /> should also work iirc. For small images, performance should be okay. If you need hundreds of large images pulled over network, one may want to consider using async I/O operations, but not in this use-case imo.
